src/tagmanager/tm_query.c

```diff
@@ -0,0 +1,314 @@
+/*
+*
+*   Copyright (c) 2016 Thomas Martitz <kugel@rockbox.org>
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License.
+*
+*/
+
+
+/**
+ * @file tm_query.h
+ * The TMQuery structure and methods to query tags from the global workspace.
+ */
+
+#include <glib.h>
+#include <glib-object.h>
+
+#include "utils.h"
+
+#include "tm_workspace.h"
+#include "tm_query.h"
+#include "tm_tag.h"
+#include "tm_parser.h"
+
+
+struct TMQuery
+{
+	const TMWorkspace *workspace;
+	gint               data_sources;
+	GPtrArray         *names;
+	GPtrArray         *scopes;
+	gint               type;
+	GArray            *langs;
+	gint               refcount;
+};
+
+
+static TMQuery *tm_query_dup(TMQuery *q)
+{
+	g_atomic_int_inc(&q->refcount);
+
+	g_return_val_if_fail(NULL != q, NULL);
+
+	return q;
+}
+
+
+/** Gets the GBoxed-derived GType for TMQuery
+ *
+ * @return TMQuery type. */
+GEANY_API_SYMBOL
+GType tm_query_get_type(void);
+
+G_DEFINE_BOXED_TYPE(TMQuery, tm_query, tm_query_dup, tm_query_free);
+
+static void free_g_string(gpointer data)
+{
+	g_string_free((GString *) data, TRUE);
+}
+
+
+/** Create a tag query
+ *
+ * The query can be used to retrieve tags from the tagmanager. You can
+ * optionally add filters and execute the query with tm_query_exec().
+ * Finally it must be freed with tm_query_free().
+ *
+ * @param workspace The workspace where the tags are stored
+ * @param data_sources Bitmask of data sources. @see TMQuerySource
+ *
+ * @return The opaque query pointer.
+ */
+GEANY_API_SYMBOL
+TMQuery *tm_query_new(const TMWorkspace *workspace, gint data_sources)
+{
+	TMQuery *q = g_slice_new(TMQuery);
+
+	q->workspace = workspace;
+	q->data_sources = data_sources;
+	q->names = g_ptr_array_new();
+	q->langs = g_array_new(FALSE, FALSE, sizeof(TMParserType));
+	q->scopes = g_ptr_array_new();
+	q->type = -1;
+	q->refcount = 1;
+
+	g_ptr_array_set_free_func(q->names, free_g_string);
+	g_ptr_array_set_free_func(q->scopes, g_free);
+
+	return q;
+}
+
+
+/** Decrements the reference count of @a q
+ *
+ * If the reference point drops to 0, then @a q is freed.
+ *
+ * @param q the query to free.
+ */
+GEANY_API_SYMBOL
+void tm_query_free(TMQuery *q)
+{
+	if (NULL != q && g_atomic_int_dec_and_test(&q->refcount))
+	{
+		g_ptr_array_free(q->names, TRUE);
+		g_ptr_array_free(q->scopes, TRUE);
+		g_array_free(q->langs, TRUE);
+		g_slice_free(TMQuery, q);
+	}
+}
+
+
+/** Add name filter to a query
+ *
+ * The query results will be restricted to tags matching the name. The
+ * matching is a simple prefix matching. The length to match can be
+ * limited to allow multiple tags to match a prefix.
+ *
+ * @param q The query to operate on.
+ * @param name The name to filter.
+ * @param name_len The number of characters to use (from the beginning).
+ *
+ * @return 0 on succcess, < 0 on error.
```

Comment:
The ability to return error codes is more to make it extensible. Since I want them in the plugin API the return type should be determined now (even if it always returns the same value *now*).

While gboolean would work, I've almost always regretted using in the past. gint is so much more flexible. I can change to boolean if you truly want it, though.